src/pybind/mgr/cephadm/upgrade.py

```diff
@@ -453,11 +453,42 @@ def _do_upgrade(self):
             ):
                 return
 
+            to_upgrade = []
+            known_ok_to_stop: List[str] = []
             for d in need_upgrade:
                 assert d.daemon_type is not None
                 assert d.daemon_id is not None
                 assert d.hostname is not None
 
+                if not d.container_image_id:
+                    if d.container_image_name == target_image:
+                        logger.debug(
+                            'daemon %s has unknown container_image_id but has correct image name' % (d.name()))
+                        continue
+
+                if known_ok_to_stop:
+                    if d.name() in known_ok_to_stop:
+                        logger.info(f'Upgrade: {d.name()} is also safe to restart')
+                        to_upgrade.append(d)
+                    continue
+
+                if not self._wait_for_ok_to_stop(d, known_ok_to_stop):
+                    return
+
+                to_upgrade.append(d)
+
+                # if we don't have a list of others to consider, stop now
+                if not known_ok_to_stop:
+                    break
+
+            num = 1
+            for d in to_upgrade:
```

Comment:
you need use [`@cephadm.utils.forall_hosts`](https://github.com/ceph/ceph/blob/853e60f77e5f7288e203b7131e200350b6c94b66/src/pybind/mgr/cephadm/utils.py#L50) in order to make this in parallel.

Reply:
i'm not sure it's necessary to parallelize this tbh.. leaving that as a future optimization